app/src/main/java/io/homeassistant/companion/android/util/TLSWebViewClient.kt

```diff
@@ -0,0 +1,102 @@
+package io.homeassistant.companion.android.util
+
+import android.app.Activity
+import android.content.Context
+import android.content.ContextWrapper
+import android.os.Build
+import android.security.KeyChain
+import android.security.KeyChainAliasCallback
+import android.webkit.ClientCertRequest
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.annotation.RequiresApi
+import io.homeassistant.companion.android.common.data.keychain.KeyChainRepository
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import java.security.Principal
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import javax.inject.Inject
+
+open class TLSWebViewClient @Inject constructor(private var keyChainRepository: KeyChainRepository) : WebViewClient() {
+
+    var isTLSClientAuthNeeded = false
+        private set
+
+    var hasUserDeniedAccess = false
+        private set
+
+    private var key: PrivateKey? = null
+    private var chain: Array<X509Certificate>? = null
+
+    private fun getActivity(context: Context?): Activity? {
+        if (context == null) {
+            return null
+        } else if (context is ContextWrapper) {
+            return if (context is Activity) {
+                context
+            } else {
+                getActivity((context).baseContext)
+            }
+        }
+        return null
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        // Let the WebViewActivity know the endpoint requires TLS Client Auth
+        isTLSClientAuthNeeded = true
+
+        // Aim to obtain the private key for the whole lifecycle of the WebViewActivity
+        val activity = getActivity(view.context)
+        if (activity == null) return
+
+        runBlocking {
+            launch {
```

Comment:
This `runBlocking` and `launch` call can be removed If you remove the `suspend` modifiers from `KeyChainRepository.getPrivateKey()` and `KeyChainRepository.getCertificateChain()` as I suggested in my [other comment](https://github.com/home-assistant/android/pull/2526/files/6a606252fde6eb8330072f437082868700884a28#r890575307).

Reply:
This should be okay

Reply:
See my [comment](https://github.com/home-assistant/android/pull/2526#discussion_r901014727) on the newest changes.